src/mesh/BoundingBox.cpp

```diff
@@ -108,58 +105,67 @@ int BoundingBox::getDimension() const
 
 const std::vector<double> BoundingBox::dataVector() const
 {
-  return _bounds;
+  std::vector<double> bounds;
+  bounds.insert(bounds.end(), {_boundMin[0], _boundMax[0], _boundMin[1], _boundMax[1]});
+  if (_dimensions == 3) {
+    bounds.insert(bounds.end(), {_boundMin[2], _boundMax[2]});
+  }
+  return bounds;
 }
 
 void BoundingBox::expandBy(const BoundingBox &otherBB)
 {
+  PRECICE_ASSERT(_dimensions == otherBB.getDimension(), "Other BoundingBox with different dimensions than this bounding box cannot be used to expand it.");
   for (int d = 0; d < _dimensions; d++) {
-    _bounds[2 * d]     = std::min(_bounds[2 * d], otherBB._bounds[2 * d]);
-    _bounds[2 * d + 1] = std::max(_bounds[2 * d + 1], otherBB._bounds[2 * d + 1]);
+    _boundMin[d] = std::min(_boundMin[d], otherBB._boundMin[d]);
+    _boundMax[d] = std::max(_boundMax[d], otherBB._boundMax[d]);
   }
 }
 
 void BoundingBox::expandBy(const Vertex &vertices)
 {
-  PRECICE_ASSERT(_dimensions == vertices.getDimensions(), "Vertex with different dimensions than bounding box cannot be used to expand bounding box");
+  PRECICE_ASSERT(_dimensions == vertices.getDimensions(), "Vertex with different dimensions than this bounding box cannot be used to expand it.");
   const auto coords = vertices.rawCoords();
   for (int d = 0; d < _dimensions; ++d) {
-    _bounds.at(2 * d)     = std::min(coords[d], _bounds.at(2 * d));
-    _bounds.at(2 * d + 1) = std::max(coords[d], _bounds.at(2 * d + 1));
+    _boundMin[d] = std::min(coords[d], _boundMin[d]);
+    _boundMax[d] = std::max(coords[d], _boundMax[d]);
   }
 }
 
 void BoundingBox::expandBy(double value)
 {
-  if (empty())
+  // Adding or subtracting a constant to the default state will still make an 'illegal' box
+  if (isDefault()) {
     return;
+  }
   for (int d = 0; d < _dimensions; d++) {
-    _bounds[2 * d] -= value;
-    _bounds[2 * d + 1] += value;
+    _boundMin[d] -= value;
+    _boundMax[d] += value;
   }
 }
 
 void BoundingBox::scaleBy(double safetyFactor)
 {
-  if (empty())
+  if (isDefault()) {
     return;
+  }
   double maxSideLength = 1e-6; // we need some minimum > 0 here
   for (int d = 0; d < _dimensions; d++) {
-    if (_bounds.at(2 * d + 1) > _bounds.at(2 * d))
-      maxSideLength = std::max(maxSideLength, _bounds[2 * d + 1] - _bounds[2 * d]);
+    if (_boundMax[d] > _boundMin[d])
+      maxSideLength = std::max(maxSideLength, _boundMax[d] - _boundMin[d]);
```

Comment:
We could use Eigen functions here, right?

Reply:
Yes.
``` cpp
maxSideLength        = (_boundMax - _boundMin).maxCoeff();
```
On the other hand, what's `PRECICE_DEBUG("Merged BoundingBox {}", *this);` for in this function? We would like to also use Eigen functions to eliminate loop in the following code, but it looks like the above debug command is used as long as _boundMax and _boundMin are changed.
```cpp
  for (int d = 0; d < _dimensions; d++) {
      _boundMax[d] += safetyFactor * maxSideLength;
      _boundMin[d] -= safetyFactor * maxSideLength;
      PRECICE_DEBUG("Merged BoundingBox {}", *this);
    }
```

Reply:
`PRECICE_DEBUG` is a logging macro for additional information when debugging code. If you can get rid of the loop (i.e. move the loop into eigen), there is no need to have this information dimension-wise (it's anyway a bit duplicated).
You can safely move it behind the loop.

Reply:
Thank you. We've done it.
```cpp
void BoundingBox::scaleBy(double safetyFactor)
{
  if (!isDefault()) {
    double maxSideLength = 1e-6; // we need some minimum > 0 here
    maxSideLength        = (_boundMax - _boundMin).maxCoeff();
    _boundMax.array() += safetyFactor * maxSideLength;
    _boundMin.array() -= safetyFactor * maxSideLength;
    PRECICE_DEBUG("Merged BoundingBox {}", *this);
  }
}
```